File: Project 2/Team.cpp
// Garland Lau
// CPSC 3200-01
// 1/30/24
// P2
// Team.cpp

#include "Team.h"
#include <iostream>

Team::Team(){
    name = "No team name";
    teamEmployees =  {};
}


Team::Team(string tName, vector<Employee*> tEmployees){
    name = tName;
    teamEmployees = tEmployees;
}

Team::~Team(){
    for(auto entry : teamEmployees){
        delete entry;
    }
}

Team::Team(const Team& other){
    name = other.name;
    for(const auto& entry : other.teamEmployees){
        teamEmployees.push_back(new Employee(*entry))
    }
}

Team &Team::operator=(const Team& other) {
    if (this == &other) {
        return this;
    }
    name = other.name();
    for (auto entry: teamEmployees) {
        delete entry;
    }
    teamEmployees.clear();

    for (const auto &entry: other.teamEmployees) {
        teamEmployees.push_back(new Employee(*entry));
    }

    return *this;
}

/* Pre-Conditions: task is valid. taskSkills is not empty
 *  -Employee is valid. empSkills is not empty
 *   -Team is valid. teamEmployees is not empty
 * Post-Conditions:
 *  -calculates avg of individual employees doing a task
 *  -returns a positve number of hours
 */

int Team::estimateTeamTaskTime(Task task){
    int totalTime = 0;
    int individualTime;
    int avgTime;
    int numEmployees = teamEmployees.size();
    for(const auto& entry : teamEmployees){
        individualTime = entry->estimateTaskTime(task);
        totalTime += individualTime;
    }
    avgTime = totalTime / numEmployees;
    return avgTime;
};

/* Pre-Conditions:
 *  -Employee is valid. Employee is not already in Team
 *  -Team is valid
 * Post-Conditions:
 *  -teamEmployees is not empty. Has 1+ members
 */

void Team::addEmployee(Employee* emp){
    teamEmployees.push_back(emp);
}

/* Pre-Conditions:
 *  -Employee is valid. Employee is in Team
 *  -Team is valid. teamEmployees is not empty
 * Post-Conditions:
 *  -teamEmployees has 1 less member. Can by empty
 */
void Team::removeEmployee(string emp_name){
    for(auto it = teamEmployees.begin(); it != teamEmployees.end(); it++){
        if((*it)->getName() == emp_name){
            delete *it;
            it = teamEmployees.erase(it);
            it--;
        }
    }
}

File: Project 4/Team.cs
// Garland Lau
// CPSC 3200-01
// P5
// Team.cs

using P5;
using System;
using System.Collections.Generic;

namespace P5{

public class Team
{
/*
* Class Invariants
* 1) requires unique team names and employee names
* 2) Team has a list of Employees 
* 3) Team can add and remove employees from teams
* 4) Estimate Team Time uses all employee skills
*/    
    private string _name = "";
    private List<Employee> _teamEmployees = new List<Employee>();

    public string name
    {
        get { return _name; }
        set { _name = value; }
    }

    public List<Employee> teamEmployees
    {
        get { return _teamEmployees; }
        set
        {
            HashSet<Employee> uniqueEmployees = new HashSet<Employee>(value);
            if (uniqueEmployees.Count != value.Count)
            {
                throw new ArgumentException("Duplicate employees in teams.");
            }
            _teamEmployees = value;
            
        }
    }

    public Team(string newName, List<Employee> newTeamEmp)
    {
        if (string.IsNullOrWhiteSpace(newName))
        {
            throw new ArgumentException("Company name cannot be empty or whitespace.");
        }
        name = newName;
        teamEmployees = newTeamEmp;
    }

    public string getName() { return name; }
    public int getNumTeamMembers() { return teamEmployees.Count; }

    public int estimateTeamTime(Task task)
    {
        int totalTime = 0, avgTime = 0, individualTime = 0;
        int teamSize = teamEmployees.Count;
 
        foreach (var emp in teamEmployees)
        {
            individualTime = emp.estimateTaskTime(task);
            totalTime += individualTime;
        }
        avgTime = totalTime / teamSize;
        return avgTime;
    }

    public void addEmployee(Employee emp)
    {
        teamEmployees.Add(emp);
    }

    public void removeEmployee(string name)
    {
        teamEmployees.RemoveAll(emp => emp.name == name);
    }
}
}